pkg/util/ipvs/ipvs_linux.go

```diff
@@ -0,0 +1,459 @@
+type IPProto uint16
+
+//// goroutine to listen for D-Bus signals
+func (runner *runner) dbusSignalHandler(bus utildbus.Connection) {
+	firewalld := bus.Object(firewalldName, firewalldPath)
+
+	for s := range runner.signal {
+		if s == nil {
+			// Unregister
+			bus.Signal(runner.signal)
+			return
+		}
+
+		switch s.Name {
+		case "org.freedesktop.DBus.NameOwnerChanged":
+			name := s.Body[0].(string)
+			new_owner := s.Body[2].(string)
+
+			if name != firewalldName || len(new_owner) == 0 {
+				continue
+			}
+
+			// FirewallD startup (specifically the part where it deletes
+			// all existing iptables rules) may not yet be complete when
+			// we get this signal, so make a dummy request to it to
+			// synchronize.
+			firewalld.Call(firewalldInterface+".getDefaultZone", 0)
+
+			runner.reload()
+		case firewalldInterface + ".Reloaded":
+			runner.reload()
+		}
+	}
+}
+
+// AddReloadFunc is part of Interface
+func (runner *runner) AddReloadFunc(reloadFunc func()) {
+	runner.reloadFuncs = append(runner.reloadFuncs, reloadFunc)
+}
+
+//// runs all reload funcs to re-sync iptables rules
+func (runner *runner) reload() {
+	glog.V(1).Infof("reloading iptables rules")
+
+	for _, f := range runner.reloadFuncs {
+		f()
+	}
+}
+
+func (runner *runner) InitIpvsInterface() error {
+	glog.V(6).Infof("Preparation for ipvs")
+
+	//Connect to DBUS first
+	runner.connectToFirewallD()
+
+	var err error
+	if ipvs_handle, err = ipvs.New(""); err != nil {
+		glog.Errorf("InitIpvsInterface: Ipvs cannot be Inited. Error: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+func (runner *runner) setSystemFlagInt(sysControl string, value int) error {
+	if val, err := runner.sysctl.GetSysctl(sysControl); err == nil && val != value {
+		runner.sysctl.SetSysctl(sysControl, value)
+	} else if err != nil {
+		glog.Errorf("Error: System control flag [%s] cannot be set", sysControl)
+		return err
+	}
+	return nil
+}
+
+func (runner *runner) CheckAliasDevice(aliasDev string) error {
+	_, err := net.InterfaceByName(aliasDev)
+	if err == nil {
+		return nil
+	}
+	err = runner.CreateAliasDevice(aliasDev)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func getAllAlias(aliasDev string) ([]net.IP, error) {
+	i, err := net.InterfaceByName(aliasDev)
+	if err != nil {
+		return nil, err
+	}
+	addrs, err := i.Addrs()
+	if err != nil {
+		return nil, err
+	}
+	ips := make([]net.IP, 0)
+	for _, addr := range addrs {
+		ip, _, err := net.ParseCIDR(addr.String())
+		if err != nil {
+			return nil, err
+		}
+		ips = append(ips, ip)
+	}
+	return ips, nil
+
+}
+
+func (runner *runner) CreateAliasDevice(aliasDev string) error {
+
+	if aliasDev == AliasDevice {
+		// Generate device alias
+		args := []string{"link", "add", aliasDev, "type", "dummy"}
+		if _, err := runner.exec.Command(cmd, args...).CombinedOutput(); err != nil {
+			// "exit status 2" is returned from the above run command if the device already exists
+			if !strings.Contains(fmt.Sprintf("%v", err), "exit status 2") {
```

Comment:
@fisherxu I think we should check the actual return code instead of the output in case of the output is not English!